tls/extensions/s2n_client_supported_versions.c

```diff
@@ -113,27 +119,45 @@ static int s2n_extensions_client_supported_versions_process(struct s2n_connectio
 
         /* We ignore the client's preferred order and instead choose
          * the highest version that both client and server support. */
-        conn->actual_protocol_version = MAX(client_version, conn->actual_protocol_version);
+        actual_protocol_version = MAX(client_version, actual_protocol_version);
     }
 
-    POSIX_ENSURE(conn->client_protocol_version != s2n_unknown_protocol_version, S2N_ERR_UNKNOWN_PROTOCOL_VERSION);
-    POSIX_ENSURE(conn->actual_protocol_version != s2n_unknown_protocol_version, S2N_ERR_PROTOCOL_VERSION_UNSUPPORTED);
+    POSIX_ENSURE(client_protocol_version != s2n_unknown_protocol_version, S2N_ERR_UNKNOWN_PROTOCOL_VERSION);
+    POSIX_ENSURE(actual_protocol_version != s2n_unknown_protocol_version, S2N_ERR_PROTOCOL_VERSION_UNSUPPORTED);
+
+    *client_protocol_version_out = client_protocol_version;
+    *actual_protocol_version_out = actual_protocol_version;
 
     return S2N_SUCCESS;
 }
 
-static int s2n_client_supported_versions_recv(struct s2n_connection *conn, struct s2n_stuffer *in)
+static int s2n_client_supported_versions_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
 {
-    if (s2n_connection_get_protocol_version(conn) < S2N_TLS13) {
-        return S2N_SUCCESS;
-    }
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(extension);
+
+    uint8_t client_protocol_version = s2n_unknown_protocol_version;
+    uint8_t actual_protocol_version = s2n_unknown_protocol_version;
+    int result = s2n_supported_versions_process(conn, extension, &client_protocol_version, &actual_protocol_version);
 
-    int result = s2n_extensions_client_supported_versions_process(conn, in);
     if (result != S2N_SUCCESS) {
-        s2n_queue_reader_unsupported_protocol_version_alert(conn);
+        /* The supported versions extension isn't required for pre-TLS 1.3 connections. For
+         * increased compatibility, if a TLS 1.2 server can't determine a compatible version from
+         * the supported versions extension for any reason, the extension is ignored. In this case,
+         * protocol selection will fall back to using the client hello version.
+         */
+        if (s2n_connection_get_server_protocol_version(conn) < S2N_TLS13) {
+            return S2N_SUCCESS;
+        }
+
+        POSIX_GUARD(s2n_queue_reader_unsupported_protocol_version_alert(conn));
         POSIX_ENSURE(s2n_errno != S2N_ERR_SAFETY, S2N_ERR_BAD_MESSAGE);
     }
     POSIX_GUARD(result);
+
+    conn->client_protocol_version = client_protocol_version;
+    conn->actual_protocol_version = actual_protocol_version;
+
```

Comment:
Isn't this a behavior change? You're setting the actual_protocol_version based on the extension even if we're doing TLS1.2. Before, we'd ignore the extension.

What you're doing is definitely compliant with the RFC:
> If this extension is present in the ClientHello, servers MUST NOT use
   the ClientHello.legacy_version value for version negotiation and MUST
   use only the "supported_versions" extension to determine client
   preferences.

However, I'm pretty sure this was an intentional deviation because customers requested that TLS1.3 changes (and very specifically this extension) not affect their existing behavior. Theoretically, I think this change could cause a behavior change if a client sets their legacy_version to TLS1.2 but only include "TLS1.0, TLS1.3" in the supported_versions extension. The new behavior would be more correct, but that probably won't matter to customers if it causes issues for them.

Especially given your timelines, you might want to restrict your change to client_protocol_version, add a compliance comment documenting the deviation from the RFC, and open an issue to revisit.

Reply:
Yes, this is a potential behavior change. It makes sense to avoid this for now. However, this means that there's now two client protocol versions: the actual client version that's advertised in the supported versions extension, and the client version that's assumed by the server since this version isn't used. s2n-tls uses client_protocol_version in many places under the assumption that it's the highest version that the client supports. If we change the meaning of this value, it breaks all of these assumptions.

If we decide to interpret two client versions, the assumed version and the advertised version, I think there should be two client version fields to reflect this: client_protocol_version and actual_client_protocol version, and the getter can return the actual one.

Reply:
I created an issue for the behavior change: https://github.com/aws/s2n-tls/issues/4240. For now I will close this and open a new PR for a version of this without any behavior change.